sdks/python/apache_beam/ml/transforms/tft_transforms_test.py

```diff
@@ -0,0 +1,393 @@
+      expected_output = []
+      bucket = 0
+      for x in sorted(test_inputs):
+        # Increment the bucket number when crossing the boundary
+        if (bucket < len(bucket_boundaries) and x >= bucket_boundaries[bucket]):
+          bucket += 1
+        expected_output.append(np.array([bucket]))
+
+      actual_output = (result | beam.Map(lambda x: x.x))
+      assert_that(
+          actual_output, equal_to(expected_output, equals_fn=np.array_equal))
+
+
+class ComputeAndVocabUnbatchedInputType(NamedTuple):
+  x: str
+
+
+class ComputeAndVocabBatchedInputType(NamedTuple):
+  x: List[str]
+
+
+@skip_if_tft_not_available
+class ComputeAndApplyVocabTest(unittest.TestCase):
+  def test_compute_and_apply_vocabulary_unbatched_inputs(self):
+    batch_size = 100
+    num_instances = batch_size + 1
+    input_data = [{
+        'x': '%.10i' % i,  # Front-padded to facilitate lexicographic sorting.
+    } for i in range(num_instances)]
+
+    expected_data = [{
+        'x': (len(input_data) - 1) - i, # Due to reverse lexicographic sorting.
+    } for i in range(len(input_data))]
+
+    with beam.Pipeline() as p:
+      process_handler = handlers.TFTProcessHandlerSchema()
+      actual_data = (
+          p
+          | "Create" >> beam.Create(input_data)
+          | beam.Map(lambda x: ComputeAndVocabUnbatchedInputType(**x)
+                     ).with_output_types(ComputeAndVocabUnbatchedInputType)
+          | "MLTransform" >>
+          base.MLTransform(process_handler=process_handler).with_transform(
+              tft_transforms.ComputeAndApplyVocabulary(columns=['x'])))
+      actual_data |= beam.Map(lambda x: x.as_dict())
+
+      assert_that(actual_data, equal_to(expected_data))
+
+  def test_compute_and_apply_vocabulary_batched(self):
+    batch_size = 100
+    num_instances = batch_size + 1
+    input_data = [
+        {
+            'x': ['%.10i' % i, '%.10i' % (i + 1), '%.10i' % (i + 2)],
+            # Front-padded to facilitate lexicographic sorting.
+        } for i in range(0, num_instances, 3)
+    ]
+
+    # since we have 3 elements in a single batch, multiply with 3 for
+    # each iteration i on the expected output.
+    excepted_data = [
+        np.array([(len(input_data) * 3 - 1) - i,
+                  (len(input_data) * 3 - 1) - i - 1,
+                  (len(input_data) * 3 - 1) - i - 2],
+                 dtype=np.int64)  # Front-padded to facilitate lexicographic
+        # sorting.
+        for i in range(0, len(input_data) * 3, 3)
+    ]
+
+    with beam.Pipeline() as p:
+      process_handler = handlers.TFTProcessHandlerSchema()
+      result = (
+          p
+          | "Create" >> beam.Create(input_data)
+          | beam.Map(lambda x: ComputeAndVocabBatchedInputType(**x)
+                     ).with_output_types(ComputeAndVocabBatchedInputType)
+          | "MLTransform" >>
+          base.MLTransform(process_handler=process_handler).with_transform(
```

Comment:
What does "process_handler" mean? Is it an abstraction we need to expose to the user?

Reply:
ProcessHandler handles the process logic related to a respective framework. Here TFTProcessHandler implements TFT code(which uses tft_beam.AnalyzeAndTransformDataset). Different frameworks will require different implementations. 

As of now, I think it would be complicated to the users to implement a ProcessHandler since it requires a full pass on data to determine some artifacts(such a min, max) and this requires combiners. I think it would be better if we provide implementations of ProcessHandlers(TFT, scikit-learn etc) to the users and make ProcessHandler internal. What do you think?

Reply:
Yes, I think we should make this internal rather than exposed to the user.

Reply:
I meant why are we exposing the notion of ProcessHandlers to the user. But maybe this is a larger discussion.

Reply:
we are creating some data transformations with TFT. To process data using these transformations, we need to implement few things such as preparing data, schema for TFT AnalyzeAndTransformDataset etc. All the logic related to TFT will be abstracted in the ProcessHandler (TFTProcessHandler).

Why ProcessHandler? TFT is a framework and it has its own semantics.  In the future if we want to create data processing transforms using JAX, we can create a JAXProcessHandler that handles jax related processing steps.

To differentiate between frameworks, users need to pass `process_handler`. If we are just implementing data transformations with TFT, then we could remove the param `process_handler` and use it internally